### PR TITLE
SAK-33087 exclude individual gradebook item by student

### PR DESCRIPTION
**Pull request by RyanAFinney**
Created new option to be able to excuse a recorded grade for a student
Added to student grade summary

This follows the same general process comments do to be saved, however excused grades will update grade calculations.

### CODE REVIEW

**Comment by RyanAFinney** [comment]
This is the solution for adding the option to excuse individual gradebook items written at University of Dayton.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/actions/ExcuseGradeAction.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            ArrayNode catDroppedItemsArray = mapper.createArrayNode();
+            droppedItems.stream().forEach(i -> catDroppedItemsArray.add(i));
+            result.put("categoryDroppedItems", catDroppedItemsArray);
+
+            return result.toString();
+        }
+    }
+
+    @Override
+    public ActionResponse handleEvent(final JsonNode params, final AjaxRequestTarget target) {
```
[the remark this answers is not in the document]
Comment (RyanAFinney) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/actions/ExcuseGradeAction.java]: I do not think this is something that can be changed for this method

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java [hunk 2]
```diff
@@ -262,6 +262,17 @@ public GradeDefinition getGradeDefinitionForStudentForItem(String gradebookUid,
...
+	/**
+	 *
+	 * @param gradebookUid
+	 * @param assignmentId
+	 * @param studentUid
+	 * @return
+	 * @throws GradebookNotFoundException
+	 * @throws AssessmentNotFoundException
+	 */
+	public boolean getAssignmentExcuse(String gradebookUid, Long assignmentId, String studentUid)
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java]: The intent of this method isn't clear (though return type makes it apparent). Are you getting the assignment excuse or if the assignment is excused? Method name could be improved here.

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/BaseHibernateManager.java [hunk 3]
```diff
@@ -1365,6 +1365,20 @@ public void setAssignmentScoreComment(final String gradebookUid, final Long assi
...
 	}
 
+	public boolean getAssignmentExcuse(final String gradebookUid, final Long assignmentId, final String studentUid) throws GradebookNotFoundException, AssessmentNotFoundException {
+        if(gradebookUid == null || assignmentId == null || studentUid == null){
+            throw new IllegalArgumentException("null parameter passed to getAssignmentScoreComment. Values are gradebookUid:" + gradebookUid + " assignmentId:" + assignmentId + " studentUid:"+ studentUid);
+        }
+        GradebookAssignment assignment = getAssignmentWithoutStats(gradebookUid, assignmentId);
+        AssignmentGradeRecord agr = getAssignmentGradeRecord(assignment, studentUid);
+
+        if(agr == null || agr.isExcludedFromGrade() == null){
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/BaseHibernateManager.java]: You could use BooleanUtils.toBoolean to handle the bulk of this if statement

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 4]
```diff
@@ -323,11 +323,18 @@ public Object doInHibernate(final Session session) throws HibernateException {
...
 						gradeDef.setGradeComment(commentText);
+						gradeDef.setExcused(false);
 					} else {
 						gradeDef.setDateRecorded(gradeRecord.getDateRecorded());
 						gradeDef.setGraderUid(gradeRecord.getGraderId());
 						gradeDef.setGradeComment(commentText);
 
+						if(gradeRecord.isExcludedFromGrade() == null){
+						    gradeDef.setExcused(false);
+                        }else{
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: spacing

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 5]
```diff
@@ -1099,8 +1106,14 @@ private double getTotalPointsInternal(final Gradebook gradebook, final List cate
 					extraCredit = true;
 				}
 
+                boolean excused;
+                if(gradeRec.isExcludedFromGrade() == null){
+                    excused = false;
+                }else{
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: spacing, also use BooleanUtils.toBoolean

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 6]
```diff
@@ -2209,27 +2269,39 @@ public void saveGradesAndComments(final String gradebookUid, final Long gradable
...
 					  throw new GradebookSecurityException();
 				  }
 			  }
-
 			  // Determine if the AssignmentGradeRecord needs to be updated
 			  String newGrade = StringUtils.trimToEmpty(gradeDef.getGrade());
 			  Double convertedGrade = convertInputGradeToPoints(gradebook.getGrade_type(), mapping, assignment.getPointsPossible(), newGrade);
 			  AssignmentGradeRecord gradeRec = studentIdGradeRecordMap.get(studentId);
+			  boolean currentExcuse;
+			  if (gradeRec == null || gradeRec.isExcludedFromGrade() == null) {
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: BooleanUtils.toBoolean

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 7]
```diff
@@ -2849,7 +2921,17 @@ public void applyDropScores(final Collection<AssignmentGradeRecord> gradeRecords
                     final List<AssignmentGradeRecord> gradesByCategory = new ArrayList<>();
                     final List<AssignmentGradeRecord> gradeRecordsByCatAndStudent = gradeRecordMap.get(studentId + cat.getId());
                     if(gradeRecordsByCatAndStudent != null) {
-                        gradesByCategory.addAll(gradeRecordsByCatAndStudent);
+                        for(AssignmentGradeRecord agr : gradeRecordsByCatAndStudent){
+                            boolean excluded;
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: BooleanUtils.toBoolean

Comment (RyanAFinney) [reply edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: All reviewed changes addressed in latest commit.

**Comment by RyanAFinney** [comment]
Not sure why this failed build? I just ran a clean build of these commits off master and it went without any problems.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/model/GbGradebookData.java [hunk 8]
```diff
@@ -445,6 +451,7 @@ private String getString(final String key) {
 			studentDefinition.setLastName(student.getStudentLastName());
 			studentDefinition.setHasComments(formatColumnFlags(student, g -> StringUtils.isNotBlank(g.getGradeComment())));
 			studentDefinition.setHasDroppedScores(formatColumnFlags(student, g -> g.isDroppedFromCategoryScore()));
+			studentDefinition.setHasExcuse(formatExcuseData(student));
```
Comment (plukasew) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/model/GbGradebookData.java]: You can just use formatColumnFlags(student, g -> g.isExcuse()) here and then you don't need formatExcuseData at all

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbGradeInfo.java [hunk 9]
```diff
@@ -40,6 +40,9 @@
 	@Getter
 	private final String gradeComment;
 
+	@Getter
+	private final boolean excuse;
```
Comment (plukasew) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbGradeInfo.java]: rename to "excused" for consistency in method names? ie. GbGradeInfo.isExcused(), GradeDefinition.isExcused()

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 10]
```diff
@@ -2238,18 +2288,18 @@ public void saveGradesAndComments(final String gradebookUid, final Long gradable
...
-				  if (StringUtils.isNotBlank(newGrade)) {
-					  gradeRec =  new AssignmentGradeRecord(assignment, studentId, convertedGrade);
-					  gradeRec.setGraderId(graderUid);
-					  gradeRec.setDateRecorded(gradedDate);
-					  gradeRecordsToUpdate.add(gradeRec);
-
-					  // Add a GradingEvent, which stores the actual input grade rather than the converted one
-					  final GradingEvent event = new GradingEvent(assignment, graderId, studentId, newGrade);
-					  eventsToAdd.add(event);
+				  if (StringUtils.isNotBlank(newGrade) && (gradeDef.getGrade() != null && !gradeDef.getGrade().trim().equals("") || excuse != currentExcuse)) {
```
Comment (plukasew) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: can replace gradeDef.getGrade() != null && !gradeDef.getGrade().trim().equals("") with StringUtils.isNotBlank(gradeDef.getGrade())

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 11]
```diff
@@ -323,11 +324,14 @@ public Object doInHibernate(final Session session) throws HibernateException {
 						gradeDef.setGrade(null);
 						gradeDef.setGraderUid(null);
 						gradeDef.setGradeComment(commentText);
+						gradeDef.setExcused(false);
 					} else {
 						gradeDef.setDateRecorded(gradeRecord.getDateRecorded());
 						gradeDef.setGraderUid(gradeRecord.getGraderId());
 						gradeDef.setGradeComment(commentText);
 
+						gradeDef.setExcused(BooleanUtils.toBoolean(gradeRecord.isExcludedFromGrade()));
```
Comment (plukasew) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: This BooleanUtils call could go in the setter instead and then you don't need to worry about calling every time you pass in a boolean value from an "untrusted" source

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 12]
```diff
@@ -2121,9 +2133,35 @@ public void saveGradeAndCommentForStudent(final String gradebookUid, final Long
...
+		if(gradebookUid == null || gradableObjectId == null || studentUid == null){
+			throw new IllegalArgumentException("Null gradebookUid, gradeableObjectId, or studentUid passed to saveGradeAndExcuseForStudent");
+		}
+
+		GradeDefinition gradeDef = new GradeDefinition();
+		gradeDef.setStudentUid(studentUid);
+		gradeDef.setGrade(grade);
+		gradeDef.setExcused(excuse);
+
+		List<GradeDefinition> gradeDefList = new ArrayList<GradeDefinition>();
```
Comment (plukasew) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: diamond operator

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/GradebookNgApplication.properties [hunk 13]
```diff
@@ -385,6 +385,7 @@ grade.notifications.privileges = Insufficient privileges to edit this Gradebook
 grade.notifications.revert = Revert to Previous Score
 grade.notifications.saved = All changes saved.
 grade.notifications.error = Errors were detected. See cell notifications below.
+grade.notifications.excused = Item has been excused from grade calculations.
```
Comment (plukasew) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/GradebookNgApplication.properties]: Not sure if "item" (could be understood to mean the entire column) is the best wording here. "Grade" might be more clear. There are ways for the entire column to be excluded.

**Comment by RyanAFinney** [comment]
@plukasew @steveswinsburg  Thanks for the reviewing! Made all changes mentioned.

**Comment by RyanAFinney** [comment]
@steveswinsburg I just noticed the merge conflict here-- looks like its just the imports (they're all the same). Is this something that I would fix and merge, or would that be handled by the reviewers?

**Comment by steveswinsburg** [comment]
Hi,
Merge conflicts are generally done by the developer.
Cheers
Steve

**Comment by RyanAFinney** [comment]
Merge conflicts resolved.

Thanks,

Ryan

**Comment by RyanAFinney** [comment]
Not sure why travis build failed-- it just rebuilt fine for me.

**Comment by bjones86** [comment]
I just checked out a fresh copy of this PR to test the build and received the same compile errors:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project gradebook-service-impl: Compilation failure: Compilation failure:
[ERROR] /opt/src/git/sakai/core/sakai/edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/BaseHibernateManager.java:[88,48] cannot find symbol
[ERROR]   symbol:   class EventTrackingService
[ERROR]   location: package org.sakaiproject.tool.gradebook.facades
[ERROR] /opt/src/git/sakai/core/sakai/edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/BaseHibernateManager.java:[104,15] cannot find symbol
[ERROR]   symbol:   class EventTrackingService
[ERROR]   location: class org.sakaiproject.component.gradebook.BaseHibernateManager
[ERROR] /opt/src/git/sakai/core/sakai/edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/BaseHibernateManager.java:[435,15] cannot find symbol
[ERROR]   symbol:   class EventTrackingService
[ERROR]   location: class org.sakaiproject.component.gradebook.BaseHibernateManager
[ERROR] /opt/src/git/sakai/core/sakai/edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/BaseHibernateManager.java:[439,47] cannot find symbol
[ERROR]   symbol:   class EventTrackingService
[ERROR]   location: class org.sakaiproject.component.gradebook.BaseHibernateManager
```

**Comment by RyanAFinney** [comment]
Thanks for the confirmation-- looking locally to see what is different now.

**Comment by RyanAFinney** [comment]
Should be all good now-- I'm not sure the thing Codacy flagged can be changed

**Comment by plukasew** [comment]
I don't think this method is particularly bad, but codacy is flagging it because it is a long method with a lot of branching paths. The simplest way to avoid this is probably to break the method up into smaller methods. For example, have a method that handles the excuse part, a method that handles the course grade part, etc.

**Comment by RyanAFinney** [comment]
@steveswinsburg @plukasew this has *finally* cleared all checks and made all requested changes-- thank you both for the patience-- I know its taken longer than I anticipated.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/actions/ExcuseGradeAction.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+        return new ExcuseGradeAction.ExcuseGradeResponse(
+                grade,
+                points,
+                isOverride,
+                categoryScore,
+                droppedItems);
+    }
+
+    public String getGrade(CourseGrade studentCourseGrade, GradebookPage page){
```
Comment (plukasew) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/actions/ExcuseGradeAction.java]: These 3 methods can probably be private. Other than that, looks good.

**Comment by RyanAFinney** [comment]
@plukasew changed those to private-- thanks! I didn't even think of that

**Pull request merged**